frost-secp256k1-tr/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+        let encoded_point =
+            k256::EncodedPoint::from_bytes(buf).map_err(|_| GroupError::MalformedElement)?;
+
+        match Option::<AffinePoint>::from(AffinePoint::from_encoded_point(&encoded_point)) {
+            Some(point) => {
+                if point.is_identity().into() {
+                    // This is actually impossible since the identity is encoded a a single byte
+                    // which will never happen since we receive a 33-byte buffer.
+                    // We leave the check for consistency.
+                    Err(GroupError::InvalidIdentityElement)
+                } else {
+                    Ok(ProjectivePoint::from(point))
+                }
+            }
+            None => Err(GroupError::MalformedElement),
+        }
+    }
+}
+
+fn hash_to_array(inputs: &[&[u8]]) -> [u8; 32] {
+    let mut h = Sha256::new();
+    for i in inputs {
+        h.update(i);
+    }
+    let mut output = [0u8; 32];
+    output.copy_from_slice(h.finalize().as_slice());
+    output
+}
+
+fn hash_to_scalar(domain: &[u8], msg: &[u8]) -> Scalar {
+    let mut u = [Secp256K1ScalarField::zero()];
+    hash_to_field::<ExpandMsgXmd<Sha256>, Scalar>(&[msg], &[domain], &mut u)
+        .expect("should never return error according to error cases described in ExpandMsgXmd");
+    u[0]
+}
+
+/// Context string from the ciphersuite in the [spec].
+///
+/// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-1
+const CONTEXT_STRING: &str = "FROST-secp256k1-SHA256-TR-v1";
+
+/// An implementation of the FROST(secp256k1, SHA-256) ciphersuite.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub struct Secp256K1Sha256;
+
+/// Digest the hasher to a Scalar
+pub fn hasher_to_scalar(hasher: Sha256) -> Scalar {
+    let sp = ScalarPrimitive::new(U256::from_be_slice(&hasher.finalize()))
+        .unwrap();
+    Scalar::from(&sp)
+}
+
+/// Create a BIP340 compliant tagged hash
+pub fn tagged_hash(tag: &str) -> Sha256 {
+    let mut hasher = Sha256::new();
+    let mut tag_hasher = Sha256::new();
+    tag_hasher.update(tag.as_bytes());
+    let tag_hash = tag_hasher.finalize();
+    hasher.update(tag_hash);
+    hasher.update(tag_hash);
+    hasher
+}
+
+/// Create a BIP341 compliant taproot tweak
+pub fn tweak(
+    public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
+    merkle_root: &[u8]
+) -> Scalar {
+    let mut hasher = tagged_hash("TapTweak");
+    hasher.update(public_key.to_affine().x());
+    hasher.update(merkle_root);
+    hasher_to_scalar(hasher)
+}
+
+/// Create a BIP341 compliant tweaked public key
+pub fn tweaked_public_key(
+    public_key: &<<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element,
+    merkle_root: &[u8],
+) -> <<Secp256K1Sha256 as Ciphersuite>::Group as Group>::Element {
+    let mut pk = public_key.clone();
+    if public_key.to_affine().y_is_odd().into() {
+        pk = -pk;
+    }
+    ProjectivePoint::GENERATOR * tweak(&pk, merkle_root) + pk
```

Comment:
notice that this might be the wrong tweaked public key, as it returns a `(x, y)` but in bitcoin only the x coordinate is given and so the public key used to verify the signature might be `(x, -y)`. I think you need to add

```rust
let tweaked_pubkey = ProjectivePoint::GENERATOR * tweak(&pk, merkle_root) + pk;
AffinePoint::decompact(&tweaked_pubkey.to_affine().x()).unwrap().into()
```